core/web/src/main/java/edu/uci/ics/texera/web/resource/QueryPlanResource.java

```diff
@@ -50,7 +53,68 @@
 public class QueryPlanResource {
     
     public static java.nio.file.Path resultDirectory = Utils.getTexeraHomePath().resolve("query-results");
-    
+
+    /**
+     * @param plan Logical plan to be executed
+     * @return Generic GenericWebResponse object
+     */
+    private JsonNode executeMutipleSinkPlan(MultipleSinkPlan plan)  {
+        HashMap<String, ISink> sinkMap = plan.getSinkMap();
+        ObjectNode response = new ObjectMapper().createObjectNode();
+        HashMap<String, List<Tuple>> executionResult = new HashMap<>();
+        // execute the query plan from all sink operators and collect result
+        for (HashMap.Entry<String, ISink> sinkEntry: sinkMap.entrySet()) {
+
+            ISink sinkOperator = sinkEntry.getValue();
+
+            if (sinkOperator instanceof TupleSink) {
+                TupleSink tupleSink = (TupleSink) sinkOperator;
+                tupleSink.open();
+                List<Tuple> result = tupleSink.collectAllTuples();
+                tupleSink.close();
+                executionResult.put(sinkEntry.getKey(), result);
+
+            } else if (sinkOperator instanceof VisualizationOperator) {
+                VisualizationOperator tupleSink = (VisualizationOperator) sinkOperator;
+                tupleSink.open();
+                List<Tuple> result = tupleSink.collectAllTuples();
+                tupleSink.close();
+                executionResult.put(sinkEntry.getKey(), result);
+
+            } else {
+                Engine.getEngine().evaluate(plan);
```

Comment:
@MIngjiHan 
1. I think code in `TupleSink` and `VisualizationOperator` are identical, please combine them.

2. in the else case, `Engine.getEngine().evaluate(plan);` will be problematic since it just calls `plan.getRoot()`. just do a `sink.open(); sink.processTuples(); sink.close()` will be fine.

Reply:
Yes, `TupleSink` and `VisualizationOperator` are similar in some ways.  
But they use different predicates, ``VisualizationOperator` also has some properties `SinkOperator` does not have. (for example , `chartType`).  
Combine them might be confusing.
But I think add new abstract class from two operators might be a good idea.